### ISSUE TITLE
Confusing message for invalid self type annotation

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#2950]
The error message for an accidental type annotation for the self argument is confusing.  Example:

```py
...
class A:
    def f(self):
        # type: (None) -> None
        ...
```

This gives an error like `The erased type of self 'void' is not a supertype of its class 'A'`. 

Usually a message like this would be better:

```
x.py:5:error: Invalid type for self, or extra argument type in function annotation
x.py:5:note: (Hint: typically annotations omit the type for self)
```

In Python 3, the related problem is typically a missing self:

```py
class A:
    def f(x: int) -> None: ...
```

Here a better error message would be something like this:

```
x.py:5:error: Self argument missing (or an invalid type for self)
```

Perhaps we can revert to the current message (or display it as a note) if it looks like the user was actually trying to use a self type, based on some heuristics. For example, `self: X` looks like a self type annotation since the argument is named `self`, and using a type variable `T` as the type of the first argument is potentially an attempt at a proper self type.

This was reported by a user, but I've seen it myself as well -- typically when I forget to add the self argument.
